src/ng/directive/ngClass.js

```diff
@@ -8,92 +8,93 @@
-          if (toAdd && toAdd.length) {
-            $animate.addClass(element, toAdd);
+          function addClasses(classes) {
+            var newClasses = digestClassCounts(classes, 1);
+            attr.$addClass(newClasses);
           }
-          if (toRemove && toRemove.length) {
-            $animate.removeClass(element, toRemove);
+
+          function removeClasses(classes) {
+            var newClasses = digestClassCounts(classes, -1);
+            attr.$removeClass(newClasses);
           }
-        }
 
-        function ngClassWatchAction(newVal) {
-          // eslint-disable-next-line no-bitwise
-          if (selector === true || (scope.$index & 1) === selector) {
-            var newClasses = arrayClasses(newVal || []);
-            if (!oldVal) {
-              addClasses(newClasses);
-            } else if (!equals(newVal,oldVal)) {
-              var oldClasses = arrayClasses(oldVal);
-              updateClasses(oldClasses, newClasses);
+          function digestClassCounts(classes, count) {
+            // Use createMap() to prevent class assumptions involving property
+            // names in Object.prototype
+            var classCounts = element.data('$classCounts') || createMap();
+            var classesToUpdate = [];
+            forEach(classes, function(className) {
+              if (count > 0 || classCounts[className]) {
+                classCounts[className] = (classCounts[className] || 0) + count;
+                if (classCounts[className] === +(count > 0)) {
+                  classesToUpdate.push(className);
+                }
+              }
+            });
+            element.data('$classCounts', classCounts);
+            return classesToUpdate.join(' ');
+          }
+
+          function updateClasses(oldClasses, newClasses) {
+            var toAdd = arrayDifference(newClasses, oldClasses);
+            var toRemove = arrayDifference(oldClasses, newClasses);
+            toAdd = digestClassCounts(toAdd, 1);
+            toRemove = digestClassCounts(toRemove, -1);
+            if (toAdd && toAdd.length) {
+              $animate.addClass(element, toAdd);
+            }
+            if (toRemove && toRemove.length) {
+              $animate.removeClass(element, toRemove);
             }
           }
-          if (isArray(newVal)) {
-            oldVal = newVal.map(function(v) { return shallowCopy(v); });
-          } else {
-            oldVal = shallowCopy(newVal);
+
+          function ngClassWatchAction(newVal) {
+            // eslint-disable-next-line no-bitwise
+            if (selector === true || (scope.$index & 1) === selector) {
+              var newClasses = (newVal || []).join(' ').split(' ');
+              if (!oldClasses) {
+                addClasses(newClasses);
+              } else if (!classArrayEquals(oldClasses, newClasses)) {
+                updateClasses(oldClasses, newClasses);
+              }
+              oldClasses = shallowCopy(newClasses);
```

Comment:
Trying to find a test to make it break, I found a case that breaks angular in older versions (at least since 1.3.8 -I didn't tried more-). Is the case of the first scope.$watch that you comment. 
I have the plnkr here: https://plnkr.co/edit/W1ck8dS08TCJpirWiGVA?p=preview
I solve it here, but I do not know if I should create another PR for solving it in the current code, what did you suggest?

I tried to find a failing test for the shallow copy inside the condition so I should move it. But I have found that moving the shallow copy outside the condition interferes with the $index $watch and make the new test fails.